src/evaluators/BinaryExpression.js

```diff
@@ -171,9 +171,56 @@ export function computeBinary(
   let resultType;
   const compute = () => {
     if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
-      // generate error if binary operation might throw or have side effects
-      resultType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-      return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
+      try {
+        // generate error if binary operation might throw or have side effects
+        resultType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
+        return AbstractValue.createFromBinaryOp(realm, op, lval, rval, loc);
+      } catch (x) {
+        if (x instanceof FatalError) {
+          // There is no need to revert any effects, because the above operation is pure.
+          // If this failed and one of the arguments was conditional, try each value
+          // and join the effects based on the condition.
+          if (lval instanceof AbstractValue && lval.kind === "conditional") {
+            let condition = lval.args[0];
```

Comment:
Rather than use lval.args[1] and lval.args[2] below, we usually deconstruct lval.args at this point. I.e. let "[condition, ll, lr] = lval.args;" or some such thing.